kong/tools/queue.lua

```diff
@@ -0,0 +1,369 @@
+--       retry_count      = 0,    -- number of times to retry processing
+--       batch_max_size   = 1000, -- max number of entries that handed to the process function in one invocation
+--       max_delay        = 1000, -- processing delay in milliseconds
+--     }
+--   )
+--
+--   q:add("Some value")
+--   q:add("Some other value")
+--
+--   ...
+--
+-- Given the example above,
+--
+-- * If the two `q:add()` invocations are done in quick succession, they will be passed to the
+--   handler function together in one batch.
+-- * If processing fails, it will not be retried (retry_count equals 0)
+-- * If retry_count was bigger than 0, processing would be re-queued n times before finally being discarded.
+-- * The retries are not regular: every time the handler fails, they next retry is delayed by n_try^2, up to
+--   max_retry_delay seconds.
+--
+-- The most important internal attributes of Queue are:
+--
+-- * `self.queue`: This is an array of entries, which are awaiting processing/consumption
+--
+-- Each entry has the following structure:
+-- {
+--   data = "some data",     -- data to process
+--   config = {...},         -- plugin configuration that caused the entry to be created
+--   timestamp = 12345789,   -- timestamp when the entry was queued
+-- }
+
+
+local semaphore = require "ngx.semaphore"
+
+local DEBUG = ngx.DEBUG
+local INFO = ngx.INFO
+local WARN = ngx.WARN
+local ERR = ngx.ERR
+
+
+local function now()
+  ngx.update_time()
+  return ngx.now()
+end
+
+local Queue = {}
+
+local Queue_mt = {
+  __index = Queue
+}
+
+
+---------
+-- Queue
+---------
+
+
+local queues = {}
+
+
+function Queue.exists(name)
+  return queues[name] and true or false
+end
+
+-------------------------------------------------------------------------------
+-- Initialize a queue with background retryable processing
+-- @param process function, invoked to process every payload generated
+-- @param opts table, requires `name`, optionally includes `retry_count`, `max_delay` and `batch_max_size`
+-- @return table: a Queue object.
+function Queue.get(name, handler, opts)
+
+  assert(type(name) == "string",
+    "arg #1 (name) must be a string")
+  assert(type(handler) == "function",
+    "arg #2 (handler) must be a function")
+  assert(type(opts) == "table",
+    "arg #3 (opts) must be a table")
+
+  assert(not opts.name or (opts.name == name),
+    "inconsistent queue name in argument and queue configuration")
+
+  local queue = queues[name]
+  if queue then
+    queue:log(DEBUG, "queue exists")
+    return queue
+  end
+
+  queue = {
+    name = name,
+    handler = handler,
+
+    semaphore = semaphore.new(),
+    retry_delay = 1,
+
+    running = true,
+    last_used = now(),
+    bytes_queued = 0,
+    queue = {},
+  }
+  for option, value in pairs(opts) do
+    queue[option] = value
+  end
+
+  queue = setmetatable(queue, Queue_mt)
+
+  kong.timer:named_every(name, queue.poll_time, function(premature, q)
```

Comment:
In this function, could we use upvalue `queue` directly instead of passing param `queue`?

Reply:
I think it would be possible to use the closed-over `queue` instead of passing it as a parameter, if that is what you mean.  In fact, it was like that before at some point but I decided to change it.  Now I wish I had written a comment why it was a problem.  In general, though, what would be the advantage?

Reply:
Marking as resolved due to the lack of feedback.